ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/MLMemoryManager.java

```diff
@@ -0,0 +1,164 @@
+        Preconditions.checkNotNull(input);
+        Preconditions.checkNotNull(response);
+        // additionalInfo cannot be null as flat object
+        additionalInfo = (additionalInfo == null) ? new HashMap<>() : additionalInfo;
+        try {
+            client
+                .execute(
+                    CreateInteractionAction.INSTANCE,
+                    new CreateInteractionRequest(
+                        conversationId,
+                        input,
+                        promptTemplate,
+                        response,
+                        origin,
+                        additionalInfo,
+                        parentIntId,
+                        traceNum
+                    ),
+                    actionListener
+                );
+        } catch (RuntimeException exception) {
+            actionListener.onFailure(exception);
+        }
+    }
+
+    /**
+     * Get the interactions associate with this conversation, sorted by recency
+     * @param conversationId the conversation whose interactions to get
+     * @param lastNInteraction Return how many interactions
+     * @param actionListener get all the final interactions that are not traces
+     */
+    public void getFinalInteractions(String conversationId, int lastNInteraction, ActionListener<List<Interaction>> actionListener) {
+        Preconditions.checkNotNull(conversationId);
+        Preconditions.checkArgument(lastNInteraction > 0, "lastN must be at least 1.");
+        log.info("Getting Interactions, conversationId {}, lastN {}", conversationId, lastNInteraction);
+
+        ActionListener<GetInteractionsResponse> al = ActionListener.wrap(getInteractionsResponse -> {
+            actionListener.onResponse(getInteractionsResponse.getInteractions());
+        }, e -> { actionListener.onFailure(e); });
+
+        try {
+            client.execute(GetInteractionsAction.INSTANCE, new GetInteractionsRequest(conversationId, lastNInteraction), al);
+        } catch (RuntimeException exception) {
+            actionListener.onFailure(exception);
+        }
+    }
+
+    /**
+     * Get the interactions associate with this conversation, sorted by recency
+     * @param parentInteractionId the parent interaction id whose traces to get
+     * @param actionListener get all the trace interactions that are only traces
+     */
+    public void getTraces(String parentInteractionId, ActionListener<List<Interaction>> actionListener) {
+        Preconditions.checkNotNull(parentInteractionId);
+        log.info("Getting traces for conversationId {}", parentInteractionId);
+
+        ActionListener<GetTracesResponse> al = ActionListener.wrap(getTracesResponse -> {
+            actionListener.onResponse(getTracesResponse.getTraces());
+        }, e -> { actionListener.onFailure(e); });
+
+        try {
+            client.execute(GetTracesAction.INSTANCE, new GetTracesRequest(parentInteractionId), al);
+        } catch (RuntimeException exception) {
```

Comment:
Why only catch runtime exception?

Reply:
See such logic in multiple places, please check if that's ok for only catching runtime exception.

Reply:
It should be Ok. The only reason it will catch error here is this client.execute method fails at runtime. All the methods in Transport Layer and Memory Layer have catch exceptions inside. 
To be on the safe side, I updated to Catch more general Exception e.
I remember quite some time ago, another SDE in my old team said that the catch exception should avoid using Exception because that couldn't give you enough info.